### ISSUE TITLE
`MLMM: Multi Modal Large Language Model`

### ISSUE THREAD

**Issue by xp1632** [xp1632/DFKI_working_log#75]
- Here's the summary of consulting a LLM specialist:

---

- We have an initial thought in #74 as follows:
![image](https://github.com/user-attachments/assets/265a3d7d-0454-4e7b-9c99-a0dd9f9ecf7c)
- We want an LLM model that can take `hybrid inputs` such as image and text

---

### Multi-Modal Large Language Model

- `MLLM` is what we are looking for, we use different modals such as image, video, and text as input for a generative AI model

- Here's **an important article that compares different MLLM models**:
 
https://medium.com/@cout.shubham/exploring-multimodal-large-language-models-a-step-forward-in-ai-626918c6a3ec

---

### Application Scenarios

1. `VQA: Vision Question Answer` 

- https://visualqa.org/

>  VQA is a new dataset containing open-ended questions about images. These questions require an understanding of vision, language, and commonsense knowledge to answer.

![image](https://github.com/user-attachments/assets/ed36c75d-1725-49b9-a16c-e250d0c542f6)

![image](https://github.com/user-attachments/assets/86a7cdee-43a4-4733-9d93-20ca55696f1e)

- VQA is commonly used in `Website Assistant` (help answer the questions from the user based on the view image of the current website page), `Automatic Drive` (also takes the frame from the camera to help answer driver's question or make decision)

---

**2. Benchmark for Segmentation: `Segment Anything`**

- `Segment anything`: https://segment-anything.com/
-  I talked about the thought of adding a `yellow marker` to help segment the cells from the background
- Then I was told the segmentation technique of `daily/natural objects` has been improved to the extreme by Meta:

- I‘ve uploaded a raw cell picture not in their image gallery
- 
![image](https://github.com/user-attachments/assets/21cc9bfd-10b8-4379-a7a7-927aaa1c6b15)

- And here's the result within 2s

![image](https://github.com/user-attachments/assets/bd9eb6cf-599e-4c05-ba27-948b07f4d76b)

- There's also more customized segmentation such as hover on the object and click to add interested point

![{694D2D09-6E6A-4582-AC00-0E94CCD8F9CB}](https://github.com/user-attachments/assets/65312bfa-e780-405a-b906-3e9a88db50ed)

- The SAM works as a `Promptable segmentation`
https://ai.meta.com/blog/segment-anything-foundation-model-image-segmentation/

**Comment by xp1632** [comment]
### How does the vision language model work?

- After reading this paper, you'll find the answer:
- LLaVa:  Large Language and Vision Assistant Visual Instruction Tuning
https://llava-vl.github.io/
- paper2: Improved Baselines with Visual Instruction Tuning 

---

- Before reading this paper, I know that, in general, they treat `image` as a type of `foreign language`
- the image is divided by grids and image feature patches are also tokenized in an `image encoder` just like the `text encoder` in text LLM

---

### How to produce the dataset for fine-tuning?

- In version `1.0 of LLaVa`: - paper1: Visual Instruction Tuning https://arxiv.org/pdf/2304.08485
-  they also mentioned how they built **a pipeline to generate the fine-tuning dataset** they required
-
- In `Cambrian-1`(https://cambrian-mllm.github.io/), they find out the dataset they collected from the public source or prompting to GPT doesn't meet their requirements
- -
- Before reading the paper, I know:
- They first build a small dataset and a small model, generate a slightly larger dataset, then introduce human intervention to improve the dataset then the small model, and get a 10% `golden dataset`
- Then based on this `10% golden dataset`, they generate the rest 90% dataset based on their data pipeline

---

### Except for the end-end generation, to avoid hallucination, is there the `white-box` GPT that we can also see the process in-between such as `code generation`?

- the previously mentioned MLLM are black boxed, which means we lose the information of `how` this model generates the answer
- to avoid hallucination which is quite common for generative models
- `ViperGPT` (https://viper.cs.columbia.edu/) also gives transparent reasoning of how this answer is generated in the white box

 
![viper_gpt](https://github.com/user-attachments/assets/0b014018-dca8-4379-a33a-644a3a1ab56a)

- In ViperGPT demo, we noticed that the reasoning for the answer is demonstrated with both python code and its respective visual inference

**Comment by xp1632** [comment]
### Paper to read:

1. **Segment Anything**: 
- paper: https://arxiv.org/pdf/2304.02643
- important blog: https://ai.meta.com/blog/segment-anything-foundation-model-image-segmentation/
- code: https://github.com/facebookresearch/segment-anything 
- **Questions**: 
- how did they prepare the dataset/mask?   
- what's a `promptable segmentation`? : 
- how did they deal with the segmentation prompts, especially the combination of additional markers, and semantic text

---

2. **LLaVA: Large Language and Vision Assistant Visual Instruction Tuning**:
- project: https://llava-vl.github.io/
- paper1: Visual Instruction Tuning https://arxiv.org/pdf/2304.08485
- paper2: Improved Baselines with Visual Instruction Tuning https://arxiv.org/pdf/2310.03744
- **Questions**: 
- how does the visual language model work?
- what's their standard/structure of dataset for finetuning?
- what does their proposed pipeline to generate the dataset look like?


---

3. **Cambrian(寒武纪)-1: A Fully Open, Vision-CentricExploration of Multimodal LLMs**
- project: https://cambrian-mllm.github.io/ 
- paper: https://arxiv.org/pdf/2406.16860
- dataset: https://huggingface.co/datasets/nyu-visionx/Cambrian-10M
- **Questions**: 
- what's a multimodal llm?
- how does the dataset look like to train a mllm?
- how to deal with the `not-enough-training data` problem(by generating more data based on `golden` ones)?
- how to build a `golden dataset` as the seed for generating the rest datasets?
- how do you know if the current dataset is `golden` or not?

---

4. **ViperGPT: Visual Inference via Python Execution for Reasoning**
- project: https://viper.cs.columbia.edu/
- paper: https://arxiv.org/pdf/2303.08128
- code: https://github.com/cvlab-columbia/viper
- **Questions**: 
- how did vipergpt link the python code and visual inference?
- what does their training dataset look like, is there also `code snippet` component in it?
- Answer after 1st reading: https://github.com/xp1632/Aalen_working_log/issues/3
